modules/database/database_addons.bal

```diff
@@ -0,0 +1,215 @@
+        fail error("Expected integer id but got a string or Nil!");
+    } else {
+        result = {experimentId: newExperimentId, name: experiment.name, description: experiment.description};
+
+        io:print(`New ExperimentId: ${newExperimentId}\n`); // TODO: remove
+
+        // map old experimentData ids to new (cloned) to avoid duplication
+        map<int> experimentDataMapping = {};
+
+        // clone timeline steps
+        stream<record {|int stepId;|}, sql:Error?> timelineStepResult = experimentDB->query(`SELECT stepId FROM TimelineStep WHERE experimentId=${oldExperimentId};`);
+        int[]|error|() timelineStepList = from var step in timelineStepResult
+            select step.stepId;
+        if timelineStepList !is error && timelineStepList !is () {
+            int counter = 0;
+            foreach var oldTimelineStepId in timelineStepList {
+                counter += 1;
+                _ = check cloneTimelineStepComplex(newExperimentId, oldExperimentId, oldTimelineStepId, experimentDataMapping);
+            }
+            io:println("ExperimentId " + newExperimentId.toString() + ": cloned " + counter.toString() + " steps.");
+        }
+        if result == () {
+            // this should logically never happen but is included for the compiler
+            return error("Experiment was empty after transaction comitted.");
+        } else {
+            return result;
+        }
+    }
+}
+
+# Clones an ExperimentData item. 
+#
+# Checks if item is already cloned (in mapping of old/new ids). If not the item is cloned and added to the mapping
+#
+# + experimentDataMapping - mutable mapping of old ExperimentData item ids to new experiment data ids to check if new experiment data needs to be created or not
+# + dataId - id of old ExperimentData item that is to be cloned
+# + newExperimentId - experiment id of the new (cloned) experiment
+# + oldExperimentId - experiment id of the experiment that is being cloned
+# + return - The new (cloned) ExperimentData id or the encountered error
+public isolated transactional function cloneExperimentData(map<int> experimentDataMapping, int dataId, int newExperimentId, int oldExperimentId) returns int|error {
+    // TODO: deep copy?
+    int? clonedExperimentDataId = experimentDataMapping[dataId.toString()];
+    if clonedExperimentDataId is () {
+        // not cloned yet => clone experiment data 
+        var experimentDataInsertResult = check experimentDB->execute(
+            `INSERT INTO ExperimentData (experimentId, name, version, location, type, contentType) 
+            SELECT ${newExperimentId}, name, version, location, type, contentType
+            FROM ExperimentData
+            WHERE experimentId = ${oldExperimentId} AND dataId = ${dataId};`
+        );
+        // add to mapping
+        var newExperimentDataId = experimentDataInsertResult.lastInsertId;
+        if newExperimentDataId is int {
+            experimentDataMapping[dataId.toString()] = newExperimentDataId;
```

Comment:
Right. Not sure why, but it does compile. The manipulation of the mapping (as used in clone-experiments) is not concurrently accessed, so there should be no problem in the current usage, but I'll add the lock statement.

Reply:
It may be that the mutable data is not module-level and therefore the violation not detected by the compiler.

Reply:
Weird. Maybe the compiler actually knows that the mutations only happen in one isolated function but I doubt it. Maybe it works because everything stays in the same thread? If I remember I will test this in the newest ballerina version and maybe report it as a bug (if it is one).

Reply:
Yes that may be a good idea. For now I added the lock statement to be on the safe side.